pci/src/vfio.rs

```diff
@@ -1911,14 +1962,29 @@ impl<M: GuestAddressSpace + Sync + Send> ExternalDmaMapping for VfioDmaMapping<M
         let mem = self.memory.memory();
         let guest_addr = GuestAddress(gpa);
         let user_addr = if mem.check_range(guest_addr, size as usize) {
-            mem.get_host_address(guest_addr).unwrap() as u64
+            match mem.get_host_address(guest_addr) {
+                Ok(t) => t as u64,
+                Err(e) => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        format!("unable to retrieve user address for gpa 0x{gpa:x} from guest memory region: {e}")
+                    ));
+                }
+            }
+        } else if self.mmio_regions.lock().unwrap().check_range(gpa, size) {
+            match self.mmio_regions.lock().unwrap().find_user_address(gpa) {
+                Ok(t) => t,
+                Err(e) => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        format!("unable to retrieve user address for gpa 0x{gpa:x} in mmio regions: {e}")
+                    ));
+                }
+            }
```

Comment:
You should be able to simply do: `self.mmio_regions.lock().unwrap().find_user_address(gpa)?`